apptail/storage/tail_tracker.go

```diff
@@ -0,0 +1,258 @@
+		t.Cached.Instances[instKey] = TailNode{}
+		t.formatMap("Current Status")
+	}
+	t.mux.Unlock()
+}
+
+// this is mainly used for testing since we are not exposing Cached via interface
+func (t *tracker) IsInstanceRegistered(instKey string) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	_, exist := t.Cached.Instances[instKey]
+
+	return exist
+}
+
+func (t *tracker) IsChildNodeInitialized(instKey string, childkey string) bool {
+	var exist bool
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childkey]; childNode_exist {
+			exist = childNode_exist
+		}
+	}
+	t.mux.Unlock()
+	return exist
+}
+
+func (t *tracker) InitializeChildNode(instKey string, childkey string, offSet int64) {
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childkey]; !childNode_exist {
+			tailNode[childkey] = &BoxedInt64{V: offSet}
+			t.Cached.Instances[instKey] = tailNode
+			t.formatMap("Current Status")
+		}
+	}
+	t.mux.Unlock()
+}
+
+func (t *tracker) GetFileCachedOffset(instkey string, fname string) int64 {
+	var offset int64
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instkey]; instance_exist {
+		offset = atomic.LoadInt64(&tailNode[fname].V)
+	}
+	t.mux.Unlock()
+	return offset
+}
+
+func (t *tracker) Update(instKey string, childKey string, childVal int64) {
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childKey]; childNode_exist {
+			atomic.StoreInt64(&tailNode[childKey].V, childVal)
+		}
+	}
+}
+
+func (t *tracker) CleanUp(cleanups map[string]bool) {
+	allInstances := make(map[string]bool)
+	t.mux.Lock()
+	for key := range t.Cached.Instances {
+		allInstances[key] = true
+
+	}
+	instanceToRemove := getEntryToCleanUp(allInstances, cleanups)
+	for key := range instanceToRemove {
+		delete(t.Cached.Instances, key)
+
+	}
+	t.formatMap("Cleaned up")
+	t.mux.Unlock()
+
+}
+
+func (t *tracker) Remove(key string) {
+	log.Infof("Removing the following key %s from cached instances", key)
+	t.mux.Lock()
+	delete(t.Cached.Instances, key)
+	t.mux.Unlock()
+	if err := t.Commit(); err != nil {
+		log.Fatal(err)
+
+	}
+}
+
+func (t *tracker) LoadTailers() {
+	t.mux.Lock()
+	t.storage.Load(&t.Cached)
+
+	t.formatMap("Loaded")
+	t.mux.Unlock()
+}
+
+func (t *tracker) getBuffer() ([]byte, error) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	return t.storage.Encode(t.Cached)
+
+}
+
+func (t *tracker) Commit() error {
+
+	bytes, err := t.getBuffer()
+
+	if err != nil {
+		return err
+
+	}
+
+	// enable for debuging
+	/*
+		t.mux.Lock()
+		t.formatMap("Storing")
+		t.mux.Unlock()
+	*/
+
+	writeMux := &sync.Mutex{}
+	writeMux.Lock()
+	err = t.storage.Write(bytes)
+	if err != nil {
+		return err
+
+	}
+	writeMux.Unlock()
+	return nil
+}
+
+func (t *tracker) formatMap(ops string) {
```

Comment:
What does this function do? Just output logging? If so, should this be at debug level? `formatMap` implies it's doing something more, but I can't see what.